### ISSUE TITLE
Free standing templates which declare variables report false errors and maybe can't be typechecked in general

### ISSUE THREAD

**Issue by zoechi** [dart-lang/angular_analyzer_plugin#176]
```html
<template #itemTemplate let-item="item" let-i="index"></template>
```
This is used to pass the template to `[ngTemplateOutlet]` or `ngForTemplate` and the context passed there can be made available by declaring template variables like `let-item="item"` or just `let-context` for the `$implicit` property.

**Comment by MichaelRFairhurst** [comment]
Hmm, where does the value for item come from? Does it come from how `#itemTemplate` is used, or from a previous `#item` declaration? Or is it a member of the enclosing class, defined in an outer `ngFor`?

If it depends on how `itemTemplate` is used, we may have to simply skip types of analyses on the contents. Or it may be unanalyzable, like ngFor would be if we didn't actually own it, it which case a means to opt out of the errors is the right solution.

**Comment by zoechi** [comment]
```html
<template [ngTemplateOutlet]="templateRef" [ngOutletContext]="{'$implicit': {'item': 'a', 'index': 2}, 'other': 'otherValue'}"></template>
```
As mentioned a template can be used this way also with `ngForTemplate` and a context being passed that is made available to the template by declaring variables like `let-self` would refer to the `$implicit` property and `let-x="other"` to the `other` property of the context (like `let-item="item"` in my initial post).

where `templateRef` is a reference from 
```dart
@ContentChildren('itemTemplate') TemplateRef templateRef;`
```
A working example (TS) would be https://plnkr.co/edit/JuwbvVe5sBGMI9Xm2DsS?p=preview

Hope this helps.

**Comment by MichaelRFairhurst** [comment]
Trying to think of a way to summarize what types of validation we could do here.

One option is to simply hide the error you reported originally. The variables should still be bound, but as a `dynamic`, which is correct because that's the most we know but we can still guarantee things around them.

A more complicated option is to treat templates as functions. We could validate them according to arguments that are applied to them, which means standard angular components using them such as ngFor and ngIf and deferred need to "call" the templates during analysis. Of course, the function overall can still be validated by the first option I proposed, as a minimum if its never "called."

How to detect when a template is "called" can involve multiple things as well. We probably have to hardcode how the core directives such as ngFor and ngIf call the template.

Angulardart doesn't yet support `ngOutletContext`, but that would be a fairly straightforward way to track a "called" and typecheck the "parameters".

More than that is probably more effort than worth it.

That means that treating templates as functions to be validated at "call" time (as opposed to using `dynamic` for all vars) would essentially boil down to validating `ngFor templateRef="x"` differently. And of course, that's assuming that its not too hard to track how the template was passed around -- if its an `@Input` on the directive, etc, that is very complicated.



I guess what this ticket should stand for is suppressing the error. I made a note on https://github.com/dart-lang/angular2/issues/46 that we could analyze it, when that gets implemented, we can open up new tickets for how to validate more from there.

**Comment by zoechi** [comment]
The behavior changed quite a bit with recent updates.

Currently I get

![image](https://cloud.githubusercontent.com/assets/405837/25996039/70853540-3715-11e7-9193-dcbb7626b845.png)

It would be great if such variables could be treated as dynamic until there are ways to do more concrete validation.
I can't imagine how this could work without additional constructs that allow to specify the type. 
Such templates can be used multiple times in different contexts (with different data passed), therefore difficult to derive.

**Comment by mk13** [comment]
Should be fixed with PR #348 . Reopen if problem persists.

**Comment by zoechi** [comment]
With these latest fixes I'll get rid of most false warnings 🎉 
Great progress!
